R/survey_statistics.r

```diff
@@ -516,10 +523,124 @@ survey_median <- function(
+#' if analised variable isn't normally distributed or there is huge variation in
+#' sampling probabilities of the survey design (or both) it could converge to
+#' normality only very slowly (with growing number of survey design degrees of
+#' freedom). Because of that constructing confidence intervals for population
+#' variance is in general a complicated task and can't be done with the same
+#' methods srvyr package provides for other statistics.
+#' @examples
+#' library(survey)
+#' data(api)
+#'
+#' dstrata <- apistrat %>%
+#'   as_survey_design(strata = stype, weights = pw)
+#'
+#' dstrata %>%
+#'   summarise(api99_var = survey_var(api99),
+#'             api99_sd = survey_sd(api99))
+#'
+#' dstrata %>%
+#'   group_by(awards) %>%
+#'   summarise(api00_var = survey_var(api00),
+#'             api00_sd = survey_sd(api00))
+#'
+#' # standard deviation and variance of the population variance estimator
+#' # are available with vartype argument
+#' # (but not for the population standard deviation estimator)
+#' dstrata %>%
+#'   summarise(api99_variance = survey_var(api99, vartype = c("se", "var")))
+#' @export
+
+survey_var <- function(
+  x, na.rm = FALSE, vartype = c("se", "var"), .svy = current_svy(), ...
+) {
+  UseMethod("survey_var", .svy)
+}
+
+#' @export
+survey_var.tbl_svy <- function(
+  x, na.rm = FALSE, vartype = c("se", "var"), .svy = current_svy(), ...
+) {
+  if (!is.null(vartype)) {
+    vartype <- if (missing(vartype)) "se" else match.arg(vartype, several.ok = TRUE)
+  }
+  df <- Inf
+  if (missing(x)) stop("Variable should be provided as an argument to survey_var().")
+  stop_for_factor(x)
+  if (class(x) == "logical") x <- as.integer(x)
+
+  .svy <- set_survey_vars(.svy, x)
+  stat <- survey::svyvar(~`__SRVYR_TEMP_VAR__`, .svy, na.rm = na.rm)
+
+  out <- get_var_est(stat, vartype, level = NULL, df = df, deff = FALSE)
+  out
+}
+
+#' @export
+survey_var.grouped_svy <- function(
+  x, na.rm = FALSE, vartype = c("se", "var"), .svy = current_svy(), ...
+) {
+  if (!is.null(vartype)) {
+    vartype <- if (missing(vartype)) "se" else match.arg(vartype, several.ok = TRUE)
+  }
+  df <- Inf
+  if (missing(x)) stop("survey_var() can't be used with regards to the grouping variable.")
+  stop_for_factor(x)
+  if (class(x) == "logical") x <- as.integer(x)
+  .svy <- set_survey_vars(.svy, x)
+  grps_formula <- survey::make.formula(group_vars(.svy))
+
+  if (any(dplyr::count(.svy$variables)$n < 2)) stop("Population variance can't be computed because some groups contain less than 2 observations.")
+
+  stat <- survey::svyby(
+    ~`__SRVYR_TEMP_VAR__`, grps_formula, .svy, survey::svyvar,
+    na.rm = na.rm
+  )
+  out <- get_var_est(
+    stat, vartype, grps = group_vars(.svy), level = NULL, df = df, deff = FALSE
+  )
+  out
+}
+
+#' @export
+#' @rdname survey_var
+survey_sd <- function(
+  x, na.rm = FALSE, .svy = current_svy(), ...
+) {
+  survey_var(
+     x, na.rm = na.rm, vartype = NULL, .svy = .svy
+  ) %>% mutate(
```

Comment:
Can you remove the use of the pipe (`%>%`)? It can make for weirder error messages (not that srvyr is particularly good about this anyways)